### PR TITLE
Implement daedalean-auto for CS.R.45

### PR DESCRIPTION
**Pull request by Javier-varez**

### CODE REVIEW

File: clang-tools-extra/clang-tidy/daedalean/AutoCheck.cpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+namespace daedalean {
+
+void AutoCheck::registerMatchers(MatchFinder *Finder) {
+  // We register 3 matchers separately:
+  //  - Variable declarations (which includes parameters)
+  //  - Function declarations (to check return types)
+  //  - Lambdas (which are handled separately given they have slightly different
+  //  rules).
+
+  // This matcher should not bind to implicit variables added by clnag, that is
```
Comment (finomen) [inline clang-tools-extra/clang-tidy/daedalean/AutoCheck.cpp]: type clang

File: clang-tools-extra/test/clang-tidy/checkers/daedalean-auto.cpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  iter end();
+};
+
+void function() {
+  vector<int> Vector{};
+
+  auto i = 123u;
+  // CHECK-MESSAGES: :[[@LINE-1]]:8: warning: Do not declare auto variables [daedalean-auto]
+
+  // auto MAY be used for iterator type in for loops.
```
Comment (finomen) [inline clang-tools-extra/test/clang-tidy/checkers/daedalean-auto.cpp:31]: For iterator, not in range-based loop.
for (auto it = v.begin(); it != v.end(); ++it)

**Review by finomen — CHANGES_REQUESTED** [review]

File: clang-tools-extra/test/clang-tidy/checkers/daedalean-auto.cpp (hunk 2 above)
Comment (Javier-varez) [reply clang-tools-extra/test/clang-tidy/checkers/daedalean-auto.cpp:31]: oops, you're right, I misunderstood this.

File: clang-tools-extra/clang-tidy/daedalean/AutoCheck.cpp [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+        if (CXXRecordDecl *recordDecl =
+                dyn_cast<CXXRecordDecl>(recordType->getDecl());
+            recordDecl && recordDecl->isLambda()) {
+          // It is an auto variable containing a lambda
+          return;
+        }
+      }
+
+      // If it is constrained, we allow the type (just like we would for a
+      // template or concept)
```
Comment (finomen) [inline clang-tools-extra/clang-tidy/daedalean/AutoCheck.cpp:74]: Good idea. We need update our CS too

**Review by finomen — APPROVED** [review]

File: clang-tools-extra/clang-tidy/daedalean/AutoCheck.cpp (hunk 3 above)
Comment (Javier-varez) [reply clang-tools-extra/clang-tidy/daedalean/AutoCheck.cpp:74]: Sounds good 👍🏻 I'm collecting all changes to the CS in another task, will add it there.

**Comment by Javier-varez** [comment]
Fixed rebase conflict

**Pull request merged**